n26/api.py

```diff
@@ -295,6 +304,84 @@ def _do_request(self, method: str = GET, url: str = "/", params: dict = None,
         if len(response.content) > 0:
             return response.json()
 
+    def get_encryption_key(self, public_key: str = None) -> dict:
+        """
+        Receive public encryption key for the JSON containing the encrypted PIN
+        """
+        return self._do_request(GET, BASE_URL_DE + '/api/encryption/key', params={
+            'publicKey': public_key})
+
+    def encrypt_user_pin(self, pin: str):
+        """
+        Encrypts user PIN and prepares to send it in a format specified in the documentation
+
+        :return: base64 encoded JSON containing the encrypted PIN and its encryption key
+        """
+        # generate AES256 key and IV
+        random_password = Random.get_random_bytes(32)
+        salt = Random.get_random_bytes(16)
+        key = PBKDF2(random_password, salt, 32, count=1000000, hmac_hash_module=SHA512)
+        iv = Random.new().read(AES.block_size)
+        key64 = base64.b64encode(key)
+        iv64 = base64.b64encode(iv)
+        # encode the key and iv as a json string
+        aes_secret = {
+        'secretKey': key64,
+        'iv': iv64,
+        }
+        # json string has to be represented in byte form for encryption
+        # simplejson used instead of json since regular json can handle only strings, not base64 encoded data
+        unencrypted_aes_secret = bytes(simplejson.dumps(aes_secret), 'utf-8')
```

Comment:
The problem is, if I try to simply `aes_secret.decode('utf-8')`, python throws an error that a 'dict' object has no attribute 'decode'. At the same time, neither does `json.dumps(aes_secret)` work, since an object of type 'bytes' is not JSON serializable. That was the reason behind introducing the simplejson library.

Reply:
You need to use `.decode('utf-8')` on the `bytes` objects. See this:

```
key64 = base64.b64encode(key).decode("utf-8")
iv64 = base64.b64encode(iv).decode("utf-8")
# encode the key and iv as a json string
aes_secret = {
    'secretKey': key64,
    'iv': iv64,
}
# json string has to be represented in byte form for encryption
# simplejson used instead of json since regular json can handle only strings, not base64 encoded data
unencrypted_aes_secret = bytes(json.dumps(aes_secret), 'utf-8')
```

This will do the same as `simplejson`, so we can just add those two `decode` calls and omit the simplejson dependency.